postgres_ffi/src/controlfile_utils.rs

```diff
@@ -0,0 +1,105 @@
+use crate::{ControlFileData, PG_CONTROLFILEDATA_OFFSETOF_CRC, PG_CONTROL_FILE_SIZE};
+
+use anyhow::bail;
+use bytes::{Bytes, BytesMut};
+
+// sizeof(ControlFileData)
+const SIZEOF_CONTROLDATA: usize = std::mem::size_of::<ControlFileData>();
+const OFFSETOF_CRC: usize = PG_CONTROLFILEDATA_OFFSETOF_CRC as usize;
+
+impl ControlFileData {
+    // Initialize an all-zeros ControlFileData struct
+    pub fn new() -> ControlFileData {
+        let controlfile: ControlFileData;
+
+        let b = [0u8; SIZEOF_CONTROLDATA];
+        controlfile =
+            unsafe { std::mem::transmute::<[u8; SIZEOF_CONTROLDATA], ControlFileData>(b) };
+
+        controlfile
+    }
+}
+
+///
+/// Interpret a slice of bytes as a Postgres control file.
+///
+pub fn decode_pg_control(buf: &[u8]) -> Result<ControlFileData, anyhow::Error> {
+    // Check that the slice has the expected size. The control file is
+    // padded with zeros up to a 512 byte sector size, so accept a
+    // larger size too, so that the caller can just the whole file
+    // contents without knowing the exact size of the struct.
+    if buf.len() < SIZEOF_CONTROLDATA {
+        bail!("control file is too short");
+    }
+
+    // Compute the expected CRC of the content.
+    let expectedcrc = crc32c::crc32c(&buf[0..OFFSETOF_CRC]);
+
+    // Convert the slice into an array of the right size, and use `transmute` to
+    // reinterpret the raw bytes as a ControlFileData struct.
+    //
+    // NB: Ideally we would use 'zerocopy::FromBytes' for this, but bindgen doesn't
+    // derive FromBytes for us. The safety of this depends on the same constraints
+    // as for FromBytes, namely, all of its fields must implement FromBytes. That
+    // includes the primitive integer types, like `u8`, `u16`, `u32`, `u64` and their
+    // signed variants. But `bool` is not safe, because the contents of the high bits
+    // in a rust bool are undefined. In practice, PostgreSQL uses 1 to represent
+    // true and 0 for false, which is compatible with Rust bool, but let's try not to
+    // depend on it.
+    //
+    // FIXME: ControlFileData does contain 'bool's at the moment.
+    //
+    // See https://github.com/zenithdb/zenith/issues/207 for discussion on the safety
+    // of this.
+    let mut b: [u8; SIZEOF_CONTROLDATA] = [0u8; SIZEOF_CONTROLDATA];
+    b.copy_from_slice(&buf[0..SIZEOF_CONTROLDATA]);
+    let controlfile: ControlFileData =
+        unsafe { std::mem::transmute::<[u8; SIZEOF_CONTROLDATA], ControlFileData>(b) };
+
+    // Check the CRC
+    if expectedcrc != controlfile.crc {
+        bail!(
+            "invalid CRC in control file: expected {:08X}, was {:08X}",
+            expectedcrc,
+            controlfile.crc
+        );
+    }
+
+    Ok(controlfile)
+}
+
+///
+/// Convert a struct representing a Postgres control file into raw bytes.
+///
+/// The CRC is recomputed to match the contents of the fields.
+pub fn encode_pg_control(controlfile: ControlFileData) -> Bytes {
```

Comment:
I also feel that those two free functions belong to the `impl ControlFileData` above.

Then they would become

```rust
impl ControlFileData {
    pub fn decode(buf: &[u8]) -> anyhow::Result<Self>;  // becomes ControlFileData::decode at call site
    pub fn encode(&self) -> Bytes;  // e.g. ControlFileData::new().encode()
}
```

Note how we don't have to consume ControlFileData in encode.